### ISSUE TITLE
ROWAN_PROJECTS_HOME should not be REQUIRED for a login anymore ...

### ISSUE THREAD

**Issue by dalehenrich** [ericwinger/Jade#895]
After attempting to use Jadeite to debug a stone created as part of a battery test and being denied access because ROWAN_PROJECTS_HOME, I'm thinking that this requirement is a bit too strict ... it is certainly useful to have ROWAN_PROJECTS_HOME defined, but you should be able to do an awful lot without ROWAN_PROJECTS_HOME and if push comes to shove, it is possible to define ROWAN_PROJECTS_HOME from within Jadeite ...

I would suggest that we make the changes in Oscar-3.1.1 and later  ...

**Comment by ericwinger** [comment]
I did a quick test to see what would happen if a Jadeite login was attempted without $ROWAN_PROJECTS_HOME being set.

Two things broke immediately - 

1. Trying to get a project service on Unpackaged had the side effect of sending RwProject>>existsOnDisk which needs $ROWAN_PROJECTS_HOME. (stack attacked) - Short answer, a browser couldn't be opened. 
2. The new file select and save dialog relies on $ROWAN_PROJECTS_HOME to know where to open. That's used because we don't know much about the remote file system so we start there. If the env var wasn't set, what server directory do you open?
[20220809_ROWAN_PROJECTS_HOME_not_set_stack.txt]

(https://github.com/GemTalk/Jadeite/files/9294138/20220809_ROWAN_PROJECTS_HOME_not_set_stack.txt)

**Comment by dalehenrich** [comment]
My inclination is that in the cases where a ROWAN_PROJECTS_HOME is currently required, we WARN that ROWAN_PROJECTS_HOME is not set and then prompt for a path (using file browser?) and allow the user to cancel and abandon the operation ... 

> new file select and save dialog
I'd be inclined to say that in the absence of ROWAN_PROJECTS_HOME, you open HOME, in the absence of HOME you open in GEMSTONE ... <`begin-riff`> **Perhaps** a setting would be called for to define default directory location (with FileSystemGS is is possible to define a set of named locations that could be set from prefs); **Perhaps** there are shortcuts in the file browser for HOME, ROWAN_PROJECTS_HOME, GEMSTONE, GemStone logs directory; **Perhaps**, it is possible to define the shortcuts from within the file browser ... <`end-riff`> 

As I think more it might make sense that when the load project menu is selected and  ROWAN_PROJECTS_HOME is not defined, you open in HOME 

Unpackaged project is scheduled to go away so for the time being `RowanProjectService >> setExistsOnDisk` should set exists on disk to false (on ERROR) (which I believe is that you're going to get back if ROWAN_PROJECTS_HOME was actually set ...

**Comment by dalehenrich** [comment]
My inclination is that in the cases where a ROWAN_PROJECTS_HOME is currently required, we WARN that ROWAN_PROJECTS_HOME is not set and then prompt for a path (using file browser?) and allow the user to cancel and abandon the operation ... 

> new file select and save dialog
I'd be inclined to say that in the absence of ROWAN_PROJECTS_HOME, you open HOME, in the absence of HOME you open in GEMSTONE ... <`begin-riff`> **Perhaps** a setting would be called for to define default directory location (with FileSystemGS is is possible to define a set of named locations that could be set from prefs); **Perhaps** there are shortcuts in the file browser for HOME, ROWAN_PROJECTS_HOME, GEMSTONE, GemStone logs directory; **Perhaps**, it is possible to define the shortcuts from within the file browser ... <`end-riff`> 

As I think more it might make sense that when the load project menu is selected and  ROWAN_PROJECTS_HOME is not defined, you open in HOME

**Comment by ericwinger** [comment]
One of the other checks we do on login to ensure that Rowan is loaded is: 

**(System myUserProfile objectNamed: ''Rowan'') notNil** and:[(System gemEnvironmentVariable: ''ROWAN_PROJECTS_HOME'') notNil]

Are you also proposing that we should allow logins if the Rowan class is not present?

**Comment by ericwinger** [comment]
> My inclination is that in the cases where a ROWAN_PROJECTS_HOME is currently required, we WARN that ROWAN_PROJECTS_HOME is not set and then prompt for a path (using file browser?) and allow the user to cancel and abandon the operation ...

Finding every place in Jadeite that touches Rowan code which requires $ROWAN_PROJECTS_HOME seems daunting. Not to mention, testing for the presence of the env var also involves a round-trip to the server before many actions could be taken. In the case of slow server response time, that would be a performance hit. 

I propose that, instead, we only check for $ROWAN_PROJECTS_HOME once during post login and encourage the user to set it then.  The set dialog would have sufficient warnings spelling out the dire consequences of choosing not to set it. 

We could also have a menu option in the Console to set or change it as a backup.

**Comment by dalehenrich** [comment]
the Rowan test is valid ...
> daunting task ...

Let's chat tomorrow

**Comment by dalehenrich** [comment]
Just to be clear ... my expectation is that a developer who ignores the warning that ROWAN_PROJECTS_HOME is not defined is probably planning on doing something similar to what I was doing ... bring up an extent0.rowan.dbf so that I could read code ... and in my mind being in readonly mode (which can be overly restrictive in my opinion) is valid and denying an operation that might use ROWAN_PROJECTS_HOME is justifiable ...

**Comment by ericwinger** [comment]
I was able to [make a change](https://github.com/GemTalk/RowanClientServices/commit/be7c09afa427799343fcdbb567eea08c2bebd50e) so that if ROWAN_PROJECTS_HOME is not found, exists on disk is set to false.  That lets the browser open and avoids problem [#1 in the earlier comment](https://github.com/GemTalk/Jadeite/issues/895#issuecomment-1209811436).

**Comment by LisaAlmarode** [comment]
I don't think this is critical for end-2022; @dalehenrich should perhaps comment

**Comment by dalehenrich** [comment]
... I'm not sure that setting exists on disk to false is quite the right answer, so we should talk a bit about this (it is possible to load a project from disk without using the env var ROWAN_PROJECTS_HOME), so this doesn't sound right to me ... my thought was to remove the ERROR and refusal to login if ROWAN_PROJECT_HOME would be sufficient (for the reasons I mention above) ... 

For end-2022, I think that letting the system come up without requiring ROWAN_PROJECTS_HOME would be good enough ... 

We probably should discuss at next meeting ...

**Comment by dalehenrich** [comment]
I think that the minimum requirement is that ROWAN_PROJECTS_HOME should not need to be set if one is going to use Jadeite to **VIEW** code in an existing rowan stone ... right now Jadeite refuses to allow one to log into the stone and that is over-restrictive for the early release ... if we take this step now allow a login without haveing ROWAN_PROJECTS_HOME ... BEFORE we are ready to turn it over to anyone beyond our cozy little group, then you (@LisaAlmarode) and I could determine how dangerous it is to not have it set ... and if it seems to dangerous to allow anyone to attempt to use Jadeite without it set, the restore the code that rejects the login ...
